rfcs/text/0017_url_service.md

```diff
@@ -0,0 +1,462 @@
+
+Below are all the other HTTP endpoints that URL Service will provide:
+
+| HTTP method           | Path                                | Description                                                                                                                              |
+|-----------------------|-------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------|
+| __GET__ or __POST__   | `/api/url/_redirect`                | Redirects the user using HTTP 302 (temporary) redirect. Receives URL generator ID and parameters in the POST body payload.               |
+| __GET__ or __POST__   | `/go/<slug>`                        | A very short alias for `POST /api/share/url/<slug>/_redirect` endpoint.                                                                  |
+
+
+#### The URL generator "redirect" endpoint
+
+The `/api/url/_redirect` endpoint will receive the necessary payload to
+redirect the user to some destination using a URL generator, for example:
+
+```
+POST /api/url/_redirect
+{
+  "generator": "DISCOVER_DEEP_LINKS",
+  "state": {
+    "indexPattern": "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxxx"
+  }
+}
+```
+
+or
+
+```
+GET /api/url/_redirect/DISCOVER_DEEP_LINKS?state={"indexPattern":"xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxxx"}
+```
+
+
+### Legacy endpoints
+
+Below are the legacy HTTP endpoints implemented by the `share` plugin, with a plan
+of action for each endpoint:
+
+| HTTP method           | Path                                | Description                                                                                                                              |
+|-----------------------|-------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------|
+| __ANY__               | `/goto/<slug>`                      | Legacy endpoint for redirecting short URLs. We will keep this endpoint, it will work the same as the `/go/<slug>` endpoint.              |
+| __GET__               | `/api/short_url/<slug>`             | The new `GET /api/short_url/<slug>` endpoint will return a superset of the payload that the legacy endpoint now returns.                 |
+| __POST__              | `/api/shorten_url`                  | The legacy endpoints for creating short URLs. We will remove it or deprecate this endpoint and maintain it until 8.0 major release.      |
+
+
+# Drawbacks
+
+Why should we *not* do this?
+
+- Implementation cost will be few weeks, but the code complexity and quality
+  will improve.
+- There is cost of migrating existing Kibana plugins to use the new API.
+
+
+# Alternatives
+
+We haven't considered other design alternatives.
+
+One alternative is still do the short URL improvements outlined above. But
+reconsider URL generators:
+
+- Do we need URL generators at all?
```

Comment:
I think we could avoid circular dependencies by creating smaller plugins

Reply:
@joshdover It is a good question, I was also asking this a few times since the URL generators were introduced.

In general, I believe (as most devs here) that URL generators should be used from plugin contracts (which is done already now), maybe even through some application service API.

But there is also a registry of all URL generators, I don't now how good any of these reasons are, but here is what I collected:

1. As you mention, to avoid circular dependencies. Like Anton mentioned, consider Discover app wants to build a URL to Dashboard and Dashboard app wants to build a URL to Discover.
2. Help with cases when some service needs to persist the URL generators state, and it uses multiple URL generators. When it needs to use it, it then loads the generator ID and generator state (which is an opaque object from its perspective) and uses it: to navigate or to do a migration. Otherwise those services need to implement a registry of URL generators themselves internally.
3. We wanted to combine URL generators with short URL service, where you create a short URL using generator ID and its state. This way the short URL service handles automatically the state migrations (currently URL generators provide a way for you to migrate the state you feed into the generator).
4. For external, for example Cloud, to be able to use a single endpoint for any URL generator, we need a central registry of those.